src/app/services/xml-parsers/apparatus-entries-parser.service.ts

```diff
@@ -0,0 +1,128 @@
+import { Injectable } from '@angular/core';
+import { map, shareReplay } from 'rxjs/operators';
+import { AppConfig } from '../../../app/app.config';
+import { ApparatusEntriesList, ApparatusEntry, ApparatusReading, XMLElement } from '../../models/evt-models';
+import { xpath, getOuterHTML } from '../../utils/dom-utils';
+import { arrayToMap } from '../../utils/js-utils';
+import { removeSpaces } from '../../utils/xml-utils';
+import { EditionDataService } from '../edition-data.service';
+import { GenericParserService } from './generic-parser.service';
+import { WitnessesParserService } from './witnesses-parser.service';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ApparatusEntriesParserService {
+  public readonly appEntries$ = this.editionDataService.parsedEditionSource$.pipe(
+    map((source) => this.parseAppEntriesList(source)),
+    shareReplay(1),
+  );
+
+  private appEntryTagName = 'app';
+  private lemmaTagName = 'lem';
+  private readingTagName = 'rdg';
+  private readingGroupTagName = 'rdgGrp';
+  private noteTagName = 'note';
+  private significantRdgNumber = 0;
+  private parsedWitnesses = {};
+
+  constructor(
+    private editionDataService: EditionDataService,
+    private genericParserService: GenericParserService,
+    private witParserService: WitnessesParserService,
+  ) {
+  }
+
+  private parseAppEntriesList(document: XMLElement): ApparatusEntriesList {
+    const appEntries = Array.from(document.querySelectorAll<XMLElement>(this.appEntryTagName));
+
+    return this.parseAppEntries(appEntries);
+  }
+
+  private parseAppEntries(appEntries: XMLElement[]) {
+    this.witParserService.witnesses$.subscribe(data => this.parsedWitnesses = data);
```

Comment:
What happens if `this.witParserService.witnesses$` does not resolve immediately? If you need parsed witnesses data to calculate variance, you should be sure that those information are ready when you use them. So you can either put the parsing of app entry inside the subscription, or handle variance as an observable that wait `this.witParserService.witnesses$` to be resolved. I probably prefer the second option

Reply:
we should try a declarative approach leveraging rxjs